feat(inputs.linux_cpu): Add CPUFreq input plugin for Linux (v3)

fabianishere opened the pull request:
This pull request adds an input plugin for monitoring Linux' CPUFreq subsystem. This pull request builds upon the work #4215 and #8342, but given that they haven't seen any activity in a while, I would like to complete the integration of this plugin into the Telegraf repository.

Currently, the plugin only exposes `cpufreq/scaling_cur_freq`, `cpufreq/scaling_min_freq`, `cpufreq/scaling_cur_freq` and `thermal_throttle/core_throttle_count`. If desirable, we could also add other fields exposed by the CPUFreq subsystem. 
**Question to the maintainers:** the `thermal_throttle/core_throttle_count` field is a bit out of place in this plugin, since it is not part of the CPUFreq subsystem in Linux. We could limit this plugin to only expose properties from the CPUFreq subsystem and have a separate plugin for `thermal_throttle` or we could emit all properties that Linux exposes per CPU (and rename the plugin to `cpulinux` for instance)?

I have optimized the implementation to reduce the number of syscalls per gather cycle. However, a limitation of this implementation is that it does not support hotplugging CPUs, though I don't expect this to affect many people.

### Required for all PRs:
- [x] Signed CLA.
- [x] Associated README.md updated.
- [ ] Has appropriate unit tests.

Resolves #2601
Resolves #4256

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+
+type CPUFreq struct {
+	Log             telegraf.Logger `toml:"-"`
+	CPUs            []CPU           `toml:"-"`
```
`CPUs` can be unexported => `cpus`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+	ID    string
+	Path  string
+	Props map[string]*os.File
+}
```
Struct and all these properties can be unexported

zak-pawel commented on plugins/inputs/cpufreq/README.md:
````diff
@@ -0,0 +1,46 @@
+
+#### Configuration
+```toml
+[[inputs.cpufreq]]
````
```
```toml
# Collects current CPU's frequencies
[[inputs.cpufreq]]
```

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func (g *CPUFreq) Description() string {
+	return "Collect current CPU's frequencies"
```
`Collects`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+		g.PathSysfs = os.Getenv("HOST_SYS")
+	} else {
+		g.PathSysfs = defaultHostSys
+	}
```
Because of `if-else`, `host_sys` paramater from config will be never taken into consideration - it will be always overwritten by content of `os.Getenv("HOST_SYS")` or by `defaultHostSys`.

I would wrap this `if-else` with:
```
if g.PathSysfs == "" {
...
}
```

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+	return nil
+}
+
+func (g *CPUFreq) DiscoverCpus() ([]CPU, error) {
```
`discoverCpus()`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+
+	cpuDirs, err := filepath.Glob(path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
+	if err != nil {
+		return nil, err
```
If `g.PathSysfs` contains path to nonexistent directory (because of rubbish in `HOST_SYS` env or in `host_sys` parameter in config file), `cpuDirs` will be empty forever. 

User will not get any metrics and will not get any information why these metrics are not here...

I would fail right here and inform user what is wrong. You can also check: `if len(cpuDirs) == 0` and return an error containing content of this command: `path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*")`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+
+func (g *CPUFreq) Init() error {
+	if os.Getenv("HOST_SYS") != "" {
+		g.PathSysfs = os.Getenv("HOST_SYS")
```
Setting `g.PathSysfs` based on `HOST_SYS` env is undocumented and may lead to problems with not getting any metrics.

Personally, I would remove this `if` and use `defaultHostSys` (only when `host_sys` was not set in config).

fabianishere replied:
Agreed, will remove the code.

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+
+		if _, err := os.Stat(filepath.Join(dir, "cpufreq")); os.IsNotExist(err) {
+			g.Log.Warn("CPUFreq subsystem not available")
+			continue
```
Is it possible that CPUFreq subsystem is not available only for some CPUs and is available for rest of them?

If not, I would return error instead of logging a warning (fail fast ;)).

fabianishere replied:
I don't think this is possible in Linux, but I'll check to make sure.

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,157 @@
+		for prop, name := range props {
+			fd, err := os.Open(filepath.Join(dir, prop))
+			if err != nil {
+				g.Log.Warn("Failed to load property of cpu ", cpuNum, ": ", filepath.Join(dir, prop))
```
Can you add `err` to warning message?

Moreover, logging `cpuNum` seems to be redundant - it will be contained in `filepath.Join(dir, prop)`

zak-pawel reviewed (changes requested):
Please, check few minor comments.
It would be great to see a few unit tests for this plugin.

fabianishere commented:
@zak-pawel Thanks for the comments. I have tried to address them.

Do you have any advice regarding the following question:
> the thermal_throttle/core_throttle_count field is a bit out of place in this plugin, since it is not part of the CPUFreq subsystem in Linux. We could limit this plugin to only expose properties from the CPUFreq subsystem and have a separate plugin for thermal_throttle or we could emit all properties that Linux exposes per CPU (and rename the plugin to cpulinux for instance)?

I might be interested in tracking other CPU-related metrics exposed by Linux but that are not part of the CPUFreq subsystem (like `thermal_throttle/core_throttle_count`)? Do you prefer to have separate input plugins for each of the CPU subsystems or  would a single input for all per-CPU metrics exposed by Linux be a better approach for Telegraf?

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+		cpu := cpu{
+			id:    cpuNum,
+			path:  dir,
+			props: make(map[string]*os.File, 0),
```
should use `make(map[string]*os.File)` instead

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+			props: make(map[string]*os.File, 0),
+		}
+
+		props := make(map[string]string, 0)
```
should use `make(map[string]string)` instead

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+var sampleConfig = `
+  ## path for sysfs filesystem.
```
`Path`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+  ## Defaults:
+  # host_sys = "/sys"
+
+  ## Gather cpu throttles per core
```
`CPU`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	if len(cpuDirs) == 0 {
+		return nil, fmt.Errorf("no CPUs detected at: %s", path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*"))
```
`path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*")` can be saved to `var`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+		for prop, name := range props {
+			fd, err := os.Open(filepath.Join(dir, prop))
+			if err != nil {
+				g.Log.Warn("Failed to load property: ", filepath.Join(dir, prop), err)
```
`Warnf`

zak-pawel commented on plugins/inputs/cpufreq/cpufreq.go:
```diff
@@ -0,0 +1,158 @@
+		cpuNum := strings.TrimPrefix(cpuName, "cpu")
+
+		if _, err := os.Stat(filepath.Join(dir, "cpufreq")); os.IsNotExist(err) {
+			return nil, fmt.Errorf("CPUFreq subsystem not available for cpu %s", cpuNum)
```
Maybe you can write something about this subsystem (and how to enable it) in `README.md`?

Please see `README.md` for `intel_powerstat` plugin.

fabianishere replied:
CPUFreq must be enabled during build time of the kernel, though it is enabled in all mainstream distributions AFAIK. Is it it sufficient to say that it must be supported by the user's kernel (along with a description of the subsystem)?

zak-pawel commented on plugins/inputs/cpufreq/README.md:
```diff
@@ -0,0 +1,47 @@
+> cpufreq,cpu=4,host=Z370M-DS3H current_khz=1272475000i,max_khz=4900000000i,min_khz=800000000i 1604049750000000000
+> cpufreq,cpu=5,host=Z370M-DS3H current_khz=1374903000i,max_khz=4900000000i,min_khz=800000000i 1604049750000000000
+> cpufreq,cpu=6,host=Z370M-DS3H current_khz=1355753000i,max_khz=4900000000i,min_khz=800000000i 1604049750000000000
+> cpufreq,cpu=7,host=Z370M-DS3H current_khz=1153656000i,max_khz=4900000000i,min_khz=800000000i 1604049750000000000
```
Cut some `0`s to have kHz ;)

zak-pawel commented:
> @zak-pawel Thanks for the comments. I have tried to address them.
> 
> Do you have any advice regarding the following question:
> 
> > the thermal_throttle/core_throttle_count field is a bit out of place in this plugin, since it is not part of the CPUFreq subsystem in Linux. We could limit this plugin to only expose properties from the CPUFreq subsystem and have a separate plugin for thermal_throttle or we could emit all properties that Linux exposes per CPU (and rename the plugin to cpulinux for instance)?
> 
> I might be interested in tracking other CPU-related metrics exposed by Linux but that are not part of the CPUFreq subsystem (like `thermal_throttle/core_throttle_count`)? Do you prefer to have separate input plugins for each of the CPU subsystems or would a single input for all per-CPU metrics exposed by Linux be a better approach for Telegraf?

Hard to say... What metrics can be gathered from `thermal_throttle`? Is `thermal_throttle` available widely in Linux? Can it be enabled/disabled?
@ssoroka @reimda What do you think?

fabianishere commented:
@zak-pawel `thermal_throttle` is exposes on Intel systems a couple metrics that might indicate that the CPU is being thermally throttled. One of the previous authors added the `core_throttle_count` field, which counts  the amount of thermal throttle events reported by the CPU.

Perhaps to give a more concrete example of my question: Linux also exposes information about the idle states via `/sys/devices/system/cpu/cpuX/cpuidle` (e.g., C-state residency, similar to what's reported by `intel_powerstat`), which I would like to track. It would not be much work to support these metrics in this input plugin, but if we do so the name `cpufreq` for the plugin might not be appropriate.

fabianishere commented:
@zak-pawel Is there still anything I can do from my side?

zak-pawel commented:
> @zak-pawel Is there still anything I can do from my side?

I don't think so :) Just waiting for some comments from @reimda / @ssoroka about `thermal_throttle` and other subsytems you had mentioned before.

reimda commented:
Hi Fabian, thanks for picking up #4215.  It makes sense to me for this plugin to include things like /sys/devices/system/cpu/cpuX/cpuidle that are outside of the cpufreq linux subsystem.  In that case I agree a more general name would be useful to quickly communicate to the user what the plugin does.

Would the name linux_cpu be more appropriate?  There's already a linux_sysctl_fs plugin that matches this naming scheme.

cpu_linux is another option.  It has the advantage of showing up next to the cpu plugin in sorted list, so it might be more visible.  I think I prefer linux_cpu.

Both these options position this plugin as the place for linux specific metrics.  That leaves the cpu plugin as the place for cross platform metrics (through gopsutil).  It would be great to have only one cpu plugin but I don't think it's practical to expect gopsutil to provide the metrics this plugin does.

If this is going to be the place for all linux cpu metrics, I would expect the field names to match linux terms.  For example, scaling_cur_freq instead of current_khz.

fabianishere commented:
@reimda I agree with your points! My plan is to rename the plugin to `linux_cpu` and start adding other metrics that Linux exposes. I will work on this the coming weekend, so you can expect an update soon after.

fabianishere commented:
@zak-pawel I have re-requested your review. Although not all possible metrics are included yet, I want to get your opinion on the current design. I have two questions in particular:
1. When should the plugin fail? It could be the case that some metrics do not exist on the system, does that mean that we fail all together?
2. To what granularity should the user be able to control the metrics that are being collected by the plugin? I can imagine that some users are not interested in tracking all possible metrics and might want to reduce overhead and the amount of data collected.

zak-pawel commented on plugins/inputs/all/all.go:
```diff
@@ -30,6 +30,7 @@ import (
 	_ "github.com/influxdata/telegraf/plugins/inputs/couchbase"
 	_ "github.com/influxdata/telegraf/plugins/inputs/couchdb"
 	_ "github.com/influxdata/telegraf/plugins/inputs/cpu"
+	_ "github.com/influxdata/telegraf/plugins/inputs/linux_cpu"
```
Please, sort this.

zak-pawel commented on plugins/inputs/linux_cpu/README.md:
````diff
@@ -0,0 +1,62 @@
+#### Configuration
+```toml
+# Collects current CPU's frequencies
+[[inputs.cpufreq]]
````
`[[inputs.linux_cpu]]`

zak-pawel commented on plugins/inputs/linux_cpu/README.md:
````diff
@@ -0,0 +1,62 @@
+
+#### Configuration
+```toml
+# Collects current CPU's frequencies
````
`# Collects CPU metrics exposed on Linux`

zak-pawel commented on plugins/inputs/linux_cpu/README.md:
````diff
@@ -0,0 +1,62 @@
+```toml
+# Collects current CPU's frequencies
+[[inputs.cpufreq]]
+  ## Path for sysfs filesystem.
+  ## See https://www.kernel.org/doc/Documentation/filesystems/sysfs.txt
+  ## Defaults:
+  # host_sys = "/sys"
+ 
+  ## Gather CPU frequency information
+  ## Defaults:
+  # gather_cpufreq = true
+
+  ## Gather CPU throttles per core
+  ## Defaults:
+  # gather_throttles = false
````
Please sync it with content of `sampleConfig` var.

zak-pawel commented on plugins/inputs/linux_cpu/README.md:
````diff
@@ -0,0 +1,62 @@
+### Example Output
+
+```
+> linux_cpu,cpu=0,host=Z370M-DS3H scaling_cur_freq=1382522i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=1,host=Z370M-DS3H scaling_cur_freq=1094884i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=2,host=Z370M-DS3H scaling_cur_freq=1010482i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=3,host=Z370M-DS3H scaling_cur_freq=2089249i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=4,host=Z370M-DS3H scaling_cur_freq=1272475i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=5,host=Z370M-DS3H scaling_cur_freq=1374903i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=6,host=Z370M-DS3H scaling_cur_freq=1355753i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
+> linux_cpu,cpu=7,host=Z370M-DS3H scaling_cur_freq=1153656i,scaling_max_freq=4900000i,scaling_min_freq=800000i 1604049750000000000
````
Would be great to have an example with all possible fields.

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,188 @@
+type LinuxCPU struct {
+	Log             telegraf.Logger `toml:"-"`
+	PathSysfs       string          `toml:"host_sys"`
+	GatherCPUFreq   bool            `toml:"gather_cpufreq"`
+	GatherThrottles bool            `toml:"gather_throttles"`
```
Please, avoid boolean fields when a string or enumerated type would be better for future extension. Lots of boolean fields also make the code difficult to maintain.

(taken from: https://github.com/influxdata/telegraf/wiki/Review#reviewing-plugin-code )

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,188 @@
+				break
+			}
+
+			cpu.props[prop.name] = fd
```
So we may get here when `err != nil && prop.optional`. 

Do we want to get here in that case? :)

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return err
+	}
+	g.cpus = cpus
```
Theoretically, we may have `len(g.cpus) == 0`.

Shouldn't we fail fast in that case right here?

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,188 @@
+			if err != nil && !prop.optional {
+				g.Log.Warnf("Failed to load property %s: %s", filepath.Join(dir, prop.path), err)
+				failed = true
+				break
```
Don't we want to `continue` here? Is there a possibility to gather these non-optional properties which don't fail?

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,188 @@
+// +build linux
+
+package linux_cpu
```
I would love to see unit tests for this plugin.

zak-pawel reviewed (changes requested):
1. I think that plugin should fail (during `Init()`) when there are no metrics to gather at all. For other cases, it would be good to gather what is possible.
2. I think that these two flags (which should be converted to array of strings in config ;)) will control the metrics that are being collected by the plugin just fine. But I would fail as soon as possible if both of them are not enabled (== no metrics to gather).

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,211 @@
+	return strconv.ParseUint(string(buffer[:n-1]), 10, 64)
+}
+
+func contains(slice []string, str string) bool {
```
There is such function in Telegraf internals. `choice.Contains` and this can be removed :)

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,211 @@
+  ## Supported options:
+  ## "cpufreq", "thermal"
+  ## Defaults:
+  # metrics = ["cpufreq""]
```
Too many `"`

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,211 @@
+
+		var failed = false
+		for _, prop := range props {
+			fd, err := os.Open(filepath.Join(dir, prop.path))
```
You open file here and you never close it...
Better idea would be to check here if file exists and is a regular file (check `validateDbPath` function in `ras.go` for example) and to save fullPath to file in `cpu.props`. Then in `readUintFromFile` you would open, read, close every file.

fabianishere replied:
Correct. I am a bit hesitant on having the collector perform multiple syscalls (for each of these metrics) every collection. With the current approach, a file descriptor is openend for each of the metrics and performs just one read call every collection cycle. The downside of this is that the process has more open file descriptors (possibly too many).

zak-pawel reviewed (commented):
Still without unit tests :(

fabianishere commented:
Thanks @zak-pawel for the review. I have been a bit busy these weeks with other things. 

Given that I don't have a lot of experience with Go, I am still researching what's the best way to test the plugin. I am probably going to approach it the same way as the `intel_powerstat` plugin where they mocked the FS interface. Let me know if you have any suggestions yourself!

sjwang90 commented:
@fabianishere Do you have a chance to continue working on this? Or else we'll see if someone else wants to take it over.

fabianishere commented:
I was planning on finishing this at the start of next year. Currently rather busy with wrapping up some other things.

antst commented:
When run as unprivileged user, this would have no permissions to read /sys/devices/system/cpu/cpuXX/cpufreq/cpuinfo_cur_freq , but because file exists, it will keep trying to read and pollute logs.
validatePath() should also check for permissions.
Either really trying to open file, or check "o+r" flag.

fabianishere commented:
@antst Good catch! I’ll probably go with just trying to open the file.

powersj commented:
@zak-pawel I think this is ready for another review since you last looked at it? Is that true?

zak-pawel commented:
> @zak-pawel I think this is ready for another review since you last looked at it? Is that true?

I would love to see few unit tests :)

fabianishere commented:
@zak-pawel Sorry for the delay, I was quite busy with some other work last months. 

I have rebased the changes onto the master branch and added a few unit tests. Let me know if you want me to cover any additional cases.

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func validatePath(propPath string) error {
+	f, err := os.Open(propPath)
+
+	if os.IsNotExist(err) {
+		return fmt.Errorf("CPU property does not exist: [%s]", propPath)
+	}
+
+	if err != nil {
+		return fmt.Errorf("cannot get system information for CPU property: [%s] - %v", propPath, err)
+	}
+
+	_ = f.Close() // File is not written to, closing should be safe
```
Maybe this would be nicer?

```
func validatePath(propPath string) error {
	_, err := os.Lstat(propPath)
	if err != nil {
		if os.IsNotExist(err) {
			return fmt.Errorf("CPU property does not exist: [%s]", propPath)
		}
		return fmt.Errorf("cannot get system information for CPU property: [%s] - %v", propPath, err)
	}

	return nil
}
```

fabianishere replied:
This relates to the comment from @antst. Checking whether the file exists is not enough, we need to verify that it is readable as well.

zak-pawel replied:
Ok.

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,226 @@
+	}
+	defer f.Close()
+
+	buffer := make([]byte, 22)
+
+	n, err := f.Read(buffer)
+	if err != nil && err != io.EOF {
+		return 0, fmt.Errorf("error on reading file, err: %v", err)
+	} else if n == 0 {
+		return 0, nil
+	}
+
+	return strconv.ParseUint(string(buffer[:n-1]), 10, 64)
```
Maybe this would be a little bit cleaner?
```
	read, err := io.ReadAll(f)
	if err != nil {
		return 0, err
	}
	
	trimmedString := strings.TrimRight(string(read), "\n")
	return strconv.ParseUint(trimmedString, 10, 64)
```

fabianishere replied:
My preference would be the first approach, since we know that the input cannot exceed 22 bytes if the value is a `uint`, so we don't need to allocate a dynamic buffer.

zak-pawel replied:
Ok, will not insist. Do you want to return `0` as a value when file was read by was empty?

fabianishere replied:
Hmm, good question, perhaps that should also give an error. What do you think?

zak-pawel replied:
I think that missing value probably will not mean that 0 should be used as a default value instead. So error could be better option.

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu_test.go:
```diff
@@ -0,0 +1,98 @@
+	require.Error(t, plugin.Init())
+}
+
+func TestNoCPUs(t *testing.T) {
+	td := t.TempDir()
+
+	plugin := &LinuxCPU{
+		Metrics:   []string{"cpufreq"},
+		PathSysfs: td,
+	}
+	require.Error(t, plugin.Init())
+}
+
```
Would be also nice to test scenario when `discoverCpus()` returns no error and empty lists (it would be for case when directories for cpus exist but all properties cannot be loaded).

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu_test.go:
```diff
@@ -0,0 +1,98 @@
+	require.Error(t, plugin.Init())
+}
+
+func TestGatherCPUFreq(t *testing.T) {
+	td := t.TempDir()
+
+	require.NoError(t, os.MkdirAll(td+"/devices/system/cpu/cpu0/cpufreq", os.ModePerm))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq", []byte("250\n"), 0644))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu0/cpufreq/scaling_min_freq", []byte("100\n"), 0644))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu0/cpufreq/scaling_max_freq", []byte("255\n"), 0644))
+
+	require.NoError(t, os.MkdirAll(td+"/devices/system/cpu/cpu1/cpufreq", os.ModePerm))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu1/cpufreq/scaling_cur_freq", []byte("123\n"), 0644))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu1/cpufreq/scaling_min_freq", []byte("80\n"), 0644))
+	require.NoError(t, os.WriteFile(td+"/devices/system/cpu/cpu1/cpufreq/scaling_max_freq", []byte("230\n"), 0644))
```
Would be nice to test scenario when property cannot be loaded (property path doesn't exist and/or value of property cannot be parsed to uint)

zak-pawel commented on plugins/inputs/linux_cpu/linux_cpu.go:
```diff
@@ -0,0 +1,218 @@
+func init() {
+	inputs.Add("linux_cpu", func() telegraf.Input {
+		return &LinuxCPU{
+			Metrics: []string{"cpufreq", "thermal"},
```
According to READMEmd, this would be default:
`Metrics: []string{"cpufreq"},`

sspaink reviewed (approved):
Great job! Thank you so much for keeping this pull request alive, I think it looks great.

@zak-pawel is out, but it looks like all the comments have been resolved. If there are any other changes I think we should handle them in a separate PR, merging.
